### PR TITLE
feat(templates): add support for custom templates in SFDX: Create Apex Class and SFDX: Create Apex Trigger - W-23688854

### PR DESCRIPTION
**Pull request by daphne-sfdc**
### What does this PR do?
1. Adds support for custom templates in the **SFDX: Create Apex Class** and **SFDX: Create Apex Trigger** commands.
2. Grabs the built-in templates dynamically for **SFDX: Create Apex Class** instead of hard-coding the list, so the built-in templates list is up-to-date with the templates in **salesforcedx-templates**.
3. Removes the **SFDX: Create Apex Unit Test Class** command because its templates are now included in the list for **SFDX: Create Apex Class**.
4. Modifies **SFDX: Create Apex Trigger** to take user input for the **sObject** and **trigger event** prior to creation instead of putting placeholders in the newly created trigger.

### What issues does this PR fix or reference?
@W-23688854@

### Testing
**SFDX: Create Apex Class**
1. No `org-custom-metadata-templates` config variable set -> see only Built-In Templates section ✅
2. `org-custom-metadata-templates` is set to a path without any apexclass templates -> see only Built-In Templates section ✅
3. The path in `org-custom-metadata-templates` contains a template that has the same name as a built-in template -> the template is removed from the Built-In Templates list and present in the Custom Templates list, and selecting it creates an Apex class with the custom template ✅
4. The path in `org-custom-metadata-templates` contains a template that does not have the same name as a built-in template -> it's listed in the Custom Templates section and works correctly ✅
5. All 7 of the built-in templates (DefaultApexClass, ApexException, ApexUnitTest, BasicUnitTest, Batchable, InboundEmailService, Queueable) work ✅
6. Custom templates work ✅
7. All built-in templates overridden -> see only Custom Templates section ✅
8. **SFDX: Create Apex Unit Test Class** command no longer present in command palette ✅
9. **SFDX: Create Apex Unit Test Class** command no longer present in context menu when you right click on the `classes` folder in Explorer View ✅
Screenshot 2026-08-06 at 9 27 10 AM

<br>**SFDX: Create Apex Trigger**
1. No `org-custom-metadata-templates` config variable set -> template selection dropdown is skipped because the only one available is the default trigger template ✅
2. `org-custom-metadata-templates` is set to a path without any apextrigger templates -> template selection dropdown is skipped because the only one available is the default trigger template ✅
3. The path in `org-custom-metadata-templates` contains a template that has the same name as the built-in template -> the template is removed from the Built-In Templates list and present in the Custom Templates list, and selecting it creates an Apex trigger with the custom template ✅
4. The path in `org-custom-metadata-templates` contains a template that does not have the same name as a built-in template -> it's listed in the Custom Templates section and works correctly ✅
5. Built-in template works ✅
6. Custom templates work ✅
7. No Default Org Set -> sObject is a text box ✅
8. Default org present -> sObject is a dropdown with all available sObjects in the default org ✅
9. Multi-select populates the trigger correctly when 1 trigger event is selected ✅
11. Multi-select populates the trigger correctly when multiple trigger events are selected ✅
12. Cancellation works properly during both the sObject and trigger event selections ✅
Screenshot 2026-08-06 at 9 27 53 AM
Screenshot 2026-08-06 at 9 28 21 AM
Screenshot 2026-08-06 at 9 28 35 AM

### CODE REVIEW

**Comment by peternhale** [comment]
## Code Review: PR #7960 - Custom Templates Support for Apex Class & Trigger

### Overview

Excellent work adding custom template support! This PR significantly improves the developer experience by:
1. **Supporting custom templates** via `org-custom-metadata-templates` config for Apex classes and triggers
2. **Dynamically discovering built-in templates** instead of hard-coding the list
3. **Removing duplicate "Create Apex Unit Test Class" command** — now unified under "Create Apex Class"
4. **Enhancing trigger creation** with interactive sObject and event selection (instead of placeholders)

The implementation is clean, well-tested with comprehensive E2E coverage, and follows Effect-TS patterns consistently.

---

### Strengths

✅ **Dynamic template discovery** - `getBuiltInTemplateNames()` future-proofs against template additions  
✅ **Smart template override** - Custom templates with same name as built-in correctly replace them  
✅ **Excellent UX flow** - Trigger creation now collects sObject/events upfront instead of leaving placeholders  
✅ **Comprehensive E2E tests** - All scenarios covered: org vs no-org, custom vs built-in, multi-select events  
✅ **Proper cleanup** - Removed defunct command from package.json, menus, and i18n  
✅ **Type safety** - `ApexClassCreateOptions` and `ApexTriggerCreateOptions` properly widen `template` to `string`  

---

### Issues & Suggestions

#### 1. **Potential issue: Empty custom templates section**
**Location:** `createApexClass.ts:242-254`

When custom templates directory exists but has NO `.cls` files, the UI shows:
```
Built-In Templates
  [separator]
  DefaultApexClass
  ...
Custom Templates
  [separator]
  (empty - no items)
```

An empty section with just a separator looks broken.

**Suggestion:** Only add the custom separator + items if `customItems.length > 0`:

```typescript
const items: vscode.QuickPickItem[] =
  customItems.length > 0
    ? [
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_class_builtin_templates_label') },
        ...nonOverriddenBuiltInItems,
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_class_custom_templates_label') },
        ...customItems
      ]
    : nonOverriddenBuiltInItems.length > 0
      ? [
          { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_class_builtin_templates_label') },
          ...nonOverriddenBuiltInItems
        ]
      : nonOverriddenBuiltInItems;
```

Or skip the "Built-In Templates" separator entirely when there are no custom templates (cleaner):

```typescript
const items: vscode.QuickPickItem[] =
  customItems.length > 0
    ? [
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_class_builtin_templates_label') },
        ...nonOverriddenBuiltInItems,
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_class_custom_templates_label') },
        ...customItems
      ]
    : builtInItems; // No separator when showing only built-in
```

Same issue exists in `createApexTrigger.ts:387-393`.

---

#### 2. **Edge case: Trigger events joined incorrectly**
**Location:** `createApexTrigger.ts:161`

```typescript
triggerevents: [triggerevents.join(', ')],
```

This wraps the joined string in an array: `['after insert, after update']`. The template expects an array of individual events: `['after insert', 'after update']`.

**Verification needed:** Does `salesforcedx-templates` expect:
- Option A: `['after insert, after update']` (one string)
- Option B: `['after insert', 'after update']` (array of strings)

If it's Option B, change to:
```typescript
triggerevents: triggerevents,
```

**Update:** Looking at the E2E test verification on line 598:
```typescript
await expect(editorText).toContainText(`trigger ${triggerName} on Case (after insert, after update)`
```

This confirms the template produces `(after insert, after update)` — so Option A is correct. However, this is brittle: if the template format changes, the code breaks silently. Consider adding a unit test that validates the template output.

---

#### 3. **Missing error handling for custom template directory read**
**Location:** `createApexClass.ts:234`

```typescript
const entries =
  isDir && apexclassDir
    ? yield* fsService.readDirectoryWithTypes(apexclassDir).pipe(Effect.orElseSucceed(() => []))
    : [];
```

`orElseSucceed(() => [])` silently swallows ALL errors (permissions, I/O failures, etc.). Users won't know their custom templates failed to load.

**Suggestion:** Log warnings for non-ENOENT errors:

```typescript
const entries =
  isDir && apexclassDir
    ? yield* fsService.readDirectoryWithTypes(apexclassDir).pipe(
        Effect.catchAll(error => {
          if (error.code !== 'ENOENT') {
            console.warn(`Failed to read custom templates from ${apexclassDir.fsPath}:`, error);
          }
          return Effect.succeed([]);
        })
      )
    : [];
```

Same applies to `createApexTrigger.ts:371`.

---

#### 4. **Minor: Missing i18n for new trigger prompts in Japanese**
**Location:** `i18n.ja.ts:470-472`

The Japanese translation file removes several keys but doesn't add translations for the new trigger prompts:
- `apex_trigger_sobject_prompt`
- `apex_trigger_events_prompt`
- `apex_trigger_default_template_description`
- `apex_trigger_builtin_templates_label`
- `apex_trigger_custom_templates_label`

These will fall back to English for Japanese users.

**Action:** Add Japanese translations or note in PR that i18n follow-up is needed.

---

#### 5. **Potential race: Trigger E2E test assumes QuickPick timing**
**Location:** `createApexTrigger.headless.spec.ts:577-584`

```typescript
await page.keyboard.press('ArrowDown');
await page.keyboard.press('Space');
await page.keyboard.press('ArrowDown');
```

No `waitFor()` between keypresses. If the QuickPick list isn't fully rendered, keypresses may be lost or go to wrong items.

**Suggestion:** Add a small delay or wait after each navigation:

```typescript
await page.keyboard.press('ArrowDown');
await page.waitForTimeout(50);
await page.keyboard.press('Space');
```

Or verify the selected item's text after each Space press.

---

#### 6. **Type safety: `getBuiltInTemplateNames` return type**
**Location:** `templateService.ts:674-682`

```typescript
const getBuiltInTemplateNames = Effect.fn('TemplateService.getBuiltInTemplateNames')(function* (
  templateDir: string,
  filetype: RegExp
) {
  // ...
  return yield* Effect.try(() =>
    SfTemplates.CreateUtil.getCommandTemplatesForFiletype(filetype, templateDir, nodeFs, templatesRootPath)
  );
});
```

The return type is inferred from `getCommandTemplatesForFiletype`, which may return `any` or a less specific type. Explicitly type the return as `string[]`:

```typescript
const getBuiltInTemplateNames = Effect.fn('TemplateService.getBuiltInTemplateNames')(function* (
  templateDir: string,
  filetype: RegExp
): Effect.Effect<string[], unknown, never> {
  // ...
});
```

---

#### 7. **Performance: Redundant file system calls**
**Location:** `createApexClass.ts:227-234`

```typescript
const apexclassDir = customPath ? Utils.joinPath(URI.file(customPath), 'apexclass') : undefined;
const isDir = apexclassDir ? yield* fsService.isDirectory(apexclassDir) : false;
const entries =
  isDir && apexclassDir
    ? yield* fsService.readDirectoryWithTypes(apexclassDir).pipe(Effect.orElseSucceed(() => []))
    : [];
```

If `isDirectory()` returns true, then `readDirectoryWithTypes()` is guaranteed to succeed (or fail for permissions, not ENOENT). The `isDirectory` check is redundant:

```typescript
const apexclassDir = customPath ? Utils.joinPath(URI.file(customPath), 'apexclass') : undefined;
const entries = apexclassDir
  ? yield* fsService.readDirectoryWithTypes(apexclassDir).pipe(Effect.orElseSucceed(() => []))
  : [];
```

`readDirectoryWithTypes` will return `[]` via `orElseSucceed` if the directory doesn't exist, so the `isDirectory` check adds an extra syscall.

Same optimization applies to `createApexTrigger.ts`.

---

### Security & Performance

✅ **Security:** No issues - all user inputs properly validated  
✅ **Performance:** Minor optimization available (see #7), but impact is negligible  

---

### Summary

**Quality:** Very high - well-architected with excellent test coverage  
**Risk Level:** Low - all edge cases tested  
**Recommendation:** 
1. Fix empty custom template section UI (issue #1) - minor but visible to users
2. Verify trigger events array format with templates team (issue #2) - low priority if E2E confirms it works
3. Consider adding error logging for custom template read failures (issue #3) - improves debuggability

Other suggestions are optional polish. Great work! 🎉

**Comment by peternhale** [comment]
## Requested Changes

Please address the following before merge:

### 1. 🔴 Fix Empty Custom Templates Section (UI Bug)

**Files:** `createApexClass.ts:242-254`, `createApexTrigger.ts:387-393`

**Issue:** When `org-custom-metadata-templates` points to a directory with no `.cls`/`.trigger` files, the QuickPick shows an empty "Custom Templates" section (just a separator with no items below it).

**Fix:** Only show the custom templates section if `customItems.length > 0`:

**For `createApexClass.ts`:**
```typescript
const items: vscode.QuickPickItem[] =
  customItems.length > 0
    ? [
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_class_builtin_templates_label') },
        ...nonOverriddenBuiltInItems,
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_class_custom_templates_label') },
        ...customItems
      ]
    : builtInItems; // No separators when showing only built-in templates
```

**For `createApexTrigger.ts`:**
```typescript
const items: vscode.QuickPickItem[] =
  customItems.length > 0
    ? [
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_trigger_builtin_templates_label') },
        ...nonOverriddenBuiltInItems,
        { kind: vscode.QuickPickItemKind.Separator, label: nls.localize('apex_trigger_custom_templates_label') },
        ...customItems
      ]
    : [{ label: 'ApexTrigger', description: APEX_TRIGGER_TEMPLATE_DESCRIPTIONS['ApexTrigger'] }];
```

---

### 2. 🟡 Add Error Logging for Custom Template Read Failures

**Files:** `createApexClass.ts:234`, `createApexTrigger.ts:371`

**Issue:** `orElseSucceed(() => [])` silently swallows ALL errors (permissions, I/O failures). Users won't know their custom templates failed to load.

**Fix:** Log warnings for non-ENOENT errors:

```typescript
const entries =
  isDir && apexclassDir
    ? yield* fsService.readDirectoryWithTypes(apexclassDir).pipe(
        Effect.catchAll(error => {
          // Log unexpected errors (permissions, I/O failures, etc.)
          if (error.code !== 'ENOENT') {
            const channelService = yield* api.services.ChannelService;
            yield* channelService.appendToChannel(
              `Warning: Could not read custom templates from ${apexclassDir.fsPath}: ${error.message}`
            );
          }
          return Effect.succeed([]);
        })
      )
    : [];
```

Apply the same pattern to `createApexTrigger.ts`.

---

### 3. 🟡 Add Missing Japanese i18n Translations

**File:** `i18n.ja.ts`

**Issue:** New trigger prompts missing Japanese translations:
- `apex_trigger_sobject_prompt`
- `apex_trigger_events_prompt`
- `apex_trigger_default_template_description`
- `apex_trigger_builtin_templates_label`
- `apex_trigger_custom_templates_label`
- `apex_class_unit_test_template_description`
- `apex_class_basic_unit_test_template_description`
- `apex_class_batchable_template_description`
- `apex_class_queueable_template_description`
- `apex_class_builtin_templates_label`
- `apex_class_custom_templates_label`

**Fix:** Either:
- Add Japanese translations for these keys, OR
- Add a comment noting that i18n follow-up is tracked separately

---

### Priority

1. **Must fix before merge:** #1 (UI bug visible to users)
2. **Strongly recommended:** #2 (improves debuggability for custom template issues)
3. **Nice to have:** #3 (completeness, but English fallback is acceptable for now)

Please push updates addressing at least #1, ideally #1 + #2.

File: packages/salesforcedx-vscode-apex-log/src/commands/createApexTrigger.ts [hunk 1]
```diff
@@ -7,21 +7,138 @@
...
+    return yield* Effect.promise(() =>
+      vscode.window.showInputBox({ prompt: nls.localize('apex_trigger_sobject_prompt') })
+    ).pipe(
+      Effect.map(raw => raw?.trim()),
+      Effect.flatMap(promptService.considerUndefinedAsCancellation)
+    );
+  }
+
+  const items: vscode.QuickPickItem[] = sobjects
+    .filter(s => s.queryable)
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex-log/src/commands/createApexTrigger.ts]: This filters the sObject picker on `s.queryable`, but queryability and triggerability are distinct flags on the describeGlobal result (jsforce exposes both `queryable` and `triggerable` separately — see `DescribeGlobalSObjectResult`). `MetadataDescribeService.listSObjects()` currently only captures `{ name, custom, queryable }` and drops `triggerable` entirely.

Consequence: an sObject that's triggerable but not queryable would never show up here, and a queryable-but-not-triggerable object could be picked and then fail to deploy the generated trigger. Worth threading `triggerable` through `SObjectGlobalDescribeItem` and filtering on that instead — though I'd recommend confirming against a live `describeGlobal()` response first since I couldn't verify the semantics against Salesforce docs directly.

File: packages/salesforcedx-vscode-apex-log/src/commands/createApexTrigger.ts [hunk 2]
```diff
@@ -43,9 +160,9 @@ export const createApexTriggerCommand = Effect.fn('createApexTriggerCommand')(fu
     outputdir: outputDirUri,
     options: {
       triggername: triggerName,
-      triggerevents: ['before insert'],
-      sobject: 'SOBJECT',
-      template: 'ApexTrigger'
+      triggerevents: [triggerevents.join(', ')],
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex-log/src/commands/createApexTrigger.ts]: `triggerevents: [triggerevents.join(', ')]` wraps an already-formatted string into a single-element array. This works today because the vendored `@salesforce/templates` `ApexTrigger.trigger` EJS template does bare `<%= triggerEvents %>` (no `.join()`), and a one-element array's default `toString()` is just the element itself — but that's an accidental quirk of `Array.prototype.toString()`, not a documented contract.

If `@salesforce/templates` ever changes the render call, or a custom template does its own iteration expecting one event per array element, this silently breaks or produces malformed trigger event lists. Recommend passing the real array (`triggerevents` as multiple elements) so the type (`string[]`) matches the actual data — note this does change the rendered formatting slightly, from `(before insert, after update)` to `(before insert,after update)` (no space), since EJS's default array-to-string join uses a bare comma.

File: packages/salesforcedx-vscode-apex-log/src/commands/createApexClass.ts [hunk 3]
```diff
@@ -13,39 +13,72 @@ import { URI, Utils } from 'vscode-uri';
...
-  template: Schema.optional(ApexClassTemplate)
+  template: Schema.optional(Schema.String)
 });
 type CreateApexClassParams = Schema.Schema.Type<typeof CreateApexClassParams>;
 
 const promptForTemplate = Effect.fn('promptForTemplate')(function* () {
   const api = yield* (yield* ExtensionProviderService).getServicesApi;
   const promptService = yield* api.services.PromptService;
+
+  const builtInNames = yield* api.services.TemplateService.getBuiltInTemplateNames('apexclass', /.cls$/);
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex-log/src/commands/createApexClass.ts]: Unescaped regex metacharacter: `/.cls$/` should be `/\.cls$/` — the bare dot matches any character, not a literal period. Inconsistent with the correctly-escaped `/\.cls$/` two lines below (line 59) and with the trigger sibling's `/\.trigger$/`.

File: packages/salesforcedx-vscode-apex-log/src/commands/createApexClass.ts [hunk 4]
```diff
@@ -13,39 +13,72 @@ import { URI, Utils } from 'vscode-uri';
...
+  const sortedNames = ['DefaultApexClass', ...builtInNames.filter((n: string) => n !== 'DefaultApexClass')];
+  const builtInItems = sortedNames.map((label: string) => ({
+    label,
+    description: APEX_CLASS_TEMPLATE_DESCRIPTIONS[label] ?? ''
+  }));
+
+  const configService = yield* api.services.ConfigService;
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex-log/src/commands/createApexClass.ts]: This custom-template-resolution block (reading `org-custom-metadata-templates` via a hardcoded string literal, `isDirectory`/`readDirectoryWithTypes`, filtering/mapping into custom items) is duplicated almost verbatim in `createApexTrigger.ts`. `templateService.ts` already reads the same setting through the typed `OrgConfigProperties.ORG_CUSTOM_METADATA_TEMPLATES` constant (`resolveCustomTemplatesPath`) — consider exposing a shared `TemplateService` method (e.g. `getCustomTemplateNames(templateDir, filetype)`) that both commands call, instead of three independent copies of the same lookup logic that can drift.

**Review by peternhale — CHANGES_REQUESTED** [review]
Automated review findings for custom Apex template support. See inline comments for details; summary below.

1. `createApexTrigger.ts:41` — sObject picker filters on `queryable` instead of `triggerable` (plausible, worth double-checking against a real `describeGlobal()` response).
2. `createApexTrigger.ts:163` — `triggerevents` is wrapped in a bogus one-element array to work around EJS's default `Array.toString()` formatting in the vendored `@salesforce/templates` trigger template.
3. `createApexClass.ts:39` — unescaped regex `/.cls$/` should be `/\.cls$/`.
4. `createApexClass.ts:46` / `createApexTrigger.ts` — custom-template-resolution logic (config lookup + directory scan + QuickPick building) is duplicated across both command files instead of living in one shared helper (e.g. on `TemplateService`).
5. `package.nls.ja.json:4` — stale `apex_generate_unit_test_class_text` key left over after removal from `package.nls.json` / `package.json` contributes.

Note: item 5 above (stale ja i18n key) is pre-existing tech debt in package.nls.ja.json, not introduced by this PR’s diff, so I couldn’t attach it inline — flagging in this summary instead.

File: packages/salesforcedx-vscode-apex-log/src/commands/createApexTrigger.ts (hunk 2 above)
Comment (daphne-sfdc) [reply packages/salesforcedx-vscode-apex-log/src/commands/createApexTrigger.ts]: I originally had the array, and changed it because of the missing space issue - `(before insert,after update)` - as you pointed out.  Since there's a missing space, I'm leaving this one as is.

**Review by peternhale — CHANGES_REQUESTED** [review]
The earlier blocking findings are resolved. I found two remaining issues that need action before approval:

1. **Custom-template read errors are still silently suppressed** (`packages/salesforcedx-vscode-services/src/core/templateService.ts`, `getCustomTemplateNames`).

`vscode.workspace.fs.readDirectory(subdirUri)` is wrapped in `Effect.orElseSucceed(() => [])`, so permission errors and other I/O failures are indistinguishable from an empty or missing template directory. Centralizing this behavior makes the silent failure affect both Apex class and trigger creation. Preserve the empty-list behavior for an absent optional directory, but surface unexpected read failures through the extension output channel (or propagate an actionable error).

2. **New user-facing strings are untranslated in Japanese** (`packages/salesforcedx-vscode-apex-log/src/messages/i18n.ja.ts`).

This change adds class-template descriptions plus trigger prompts, template descriptions, and template-section labels in the English catalog, but none have Japanese entries. Japanese users therefore receive mixed-language QuickPick and prompt UI. Please add translations for the new keys, or remove/avoid the new strings until localization is ready.

Please address both items and rerun the relevant checks.

File: packages/salesforcedx-vscode-services/src/core/templateService.ts [hunk 5]
```diff
@@ -271,6 +292,39 @@ export class TemplateService extends Effect.Service<TemplateService>()('Template
...
+    const getCustomTemplateNames = Effect.fn('TemplateService.getCustomTemplateNames')(function* (
+      templateDir: string,
+      ext: string
+    ) {
+      const customPath = yield* resolveCustomTemplatesPath().pipe(Effect.orElseSucceed(() => undefined));
+      if (!customPath) return [];
+      const subdirUri = Utils.joinPath(URI.file(customPath), templateDir);
+      const entries = yield* Effect.tryPromise(() => vscode.workspace.fs.readDirectory(subdirUri)).pipe(
+        Effect.catchAll(e => {
+          const isNotFound = e instanceof vscode.FileSystemError && (e.code === 'FileNotFound' || e.code === 'ENOENT');
```
Comment (peternhale) [inline packages/salesforcedx-vscode-services/src/core/templateService.ts]: This `isNotFound` check can never be true, so the "preserve empty-list behavior for an absent optional directory" branch is dead code.

`vscode.workspace.fs.readDirectory(subdirUri)` is wrapped in a bare `Effect.tryPromise(() => ...)` with no `catch` option. Per Effect's docs, without a `catch`, the failure always surfaces as an `UnknownException` — the original thrown error lives on `e.error`, not on `e` itself. So `e instanceof vscode.FileSystemError` here is always `false`, and every read failure (including the completely normal case of a user's custom-templates path not having an `apexclass`/`apextrigger` subfolder) falls through to `Effect.logWarning`, which will now fire on every Apex class/trigger creation for most users instead of staying silent for that expected case.

Suggest checking `e.error instanceof vscode.FileSystemError` (or narrowing on `e.error`'s `code`) instead of `e`.

**Review by peternhale — CHANGES_REQUESTED** [review]
Checked the fixes for the previous round of feedback.

- Japanese translations (`i18n.ja.ts`): all 11 new keys are translated, looks complete ✅
- Error handling fix (`templateService.ts`): has a bug, see inline comment.

**Review by peternhale — APPROVED** [review]
All four rounds of review feedback are now addressed, including the UnknownException unwrapping fix in templateService.ts. CI is green aside from the known unrelated orgBrowserDreamhouseTestOrg scratch-org flake. Approving — please rebase to resolve the develop merge conflict before landing.
